Controller/VariantController.php

```diff
@@ -123,47 +124,142 @@ public function cgetAction(Request $request, $parentId)
     }
 
     /**
-     * Adds a new variant to this product.
+     * Adds a new variant to given product.
      *
      * @param Request $request
-     * @param $parentId
+     * @param int $parentId
      *
      * @return Response
      */
     public function postAction(Request $request, $parentId)
     {
+        $requestData = $request->request->all();
+        $userId = $this->getUser()->getId();
+        $locale = $this->getLocale($request);
+
         try {
-            $variant = $this->getManager()->addVariant($parentId, $request->get('id'), $this->getLocale($request));
+            $variant = $this->getProductVariantManager()->createVariant(
+                $parentId,
+                $requestData,
+                $locale,
+                $userId
+            );
+
+            $this->getEntityManager()->flush();
 
-            $view = $this->view($variant, 200);
+            $apiVariant = $this->getProductFactory()->createApiEntity($variant, $locale);
+
+            $view = $this->view($apiVariant, 200);
         } catch (ProductNotFoundException $exc) {
             $exception = new EntityNotFoundException($exc->getEntityName(), $exc->getId());
             $view = $this->view($exception->toArray(), 400);
+        } catch (ProductException $exc) {
+            $view = $this->view($exc->getMessage(), 400);
         }
 
         return $this->handleView($view);
     }
 
     /**
-     * Removes a variant from a product.
+     * Updates an existing variant to given product.
      *
      * @param Request $request
-     * @param $parentId
-     * @param $id
+     * @param int $parentId
+     * @param int $variantId
      *
      * @return Response
      */
-    public function deleteAction(Request $request, $parentId, $id)
+    public function putAction(Request $request, $parentId, $variantId)
     {
+        $requestData = $request->request->all();
+        $userId = $this->getUser()->getId();
+        $locale = $this->getLocale($request);
+
         try {
-            $this->getManager()->removeVariant($parentId, $id);
+            $variant = $this->getProductVariantManager()->updateVariant(
+                $variantId,
+                $requestData,
+                $locale,
+                $userId
+            );
+
+            if ($variant->getParent()->getId() !== (int) $parentId) {
+                throw new ProductException('Variant does not exists for given product.');
+            }
+
+            $this->getEntityManager()->flush();
+
+            $apiVariant = $this->getProductFactory()->createApiEntity($variant, $locale);
+
+            $view = $this->view($apiVariant, 200);
+        } catch (ProductNotFoundException $exc) {
```

Comment:
we use the fos-rest exception controller to avoid catching exceptions here.

you can omit the catch and configure the exceptions in a preprend-extension see https://github.com/sulu/sulu/blob/develop/src/Sulu/Bundle/ContentBundle/DependencyInjection/SuluContentExtension.php#L80